File: mobile/lib/data/services/wheeldeck_client.dart
import 'dart:async';
import 'dart:convert';

import 'package:stream_channel/stream_channel.dart';
import 'package:web_socket_channel/web_socket_channel.dart';

import '../../domain/models/connection_status.dart';
import '../../domain/models/connection_target.dart'
    show ConnectionTarget, defaultWheelDeckPort;
import '../../domain/models/pairing_challenge.dart';
import '../../domain/models/pairing_method.dart';
import 'dashboard_input.dart';
import 'input_mapping.dart';

/// The mobile end of the WheelDeck WebSocket protocol: dials the desktop,
/// frames `state` and `button` messages, tracks a monotonic sequence number,
/// and reports connection status. Pairing, discovery, and heartbeat wire into
/// this class from their own modules.
class WheelDeckClient {
  WheelDeckClient({
    required this.deviceId,
    Future<StreamChannel<dynamic>> Function(Uri uri)? connect,
    this.heartbeatInterval = defaultHeartbeatInterval,
    this.reconnectInterval = defaultReconnectInterval,
  }) : _connect = connect ?? _defaultConnect;

  static const int defaultPort = defaultWheelDeckPort;
  static const Duration defaultHeartbeatInterval = Duration(seconds: 2);
  static const Duration defaultReconnectInterval = Duration(seconds: 3);

  final String deviceId;
  final Duration heartbeatInterval;
  final Duration reconnectInterval;
  final Future<StreamChannel<dynamic>> Function(Uri uri) _connect;

  StreamChannel<dynamic>? _channel;
  StreamSubscription<dynamic>? _subscription;
  ConnectionStatus _status = ConnectionStatus.disconnected;
  int _seq = 0;
  String? _sessionToken;
  ConnectionTarget? _lastTarget;
  Timer? _heartbeatTimer;
  Timer? _reconnectTimer;

  void Function(ConnectionStatus status)? _onConnectionStatusChanged;
  void Function(PairingChallenge challenge)? _onPairingRequired;
  void Function(String sessionToken)? _onPairingAccepted;

  /// The most recent status, for widgets that build from state.
  ConnectionStatus get status => _status;

  /// The last connection target, used to reconnect after a lifecycle pause.
  ConnectionTarget? get lastTarget => _lastTarget;

  /// Registers the callback that receives connection status transitions.
  void onConnectionStatusChanged(
    void Function(ConnectionStatus status) callback,
  ) {
    _onConnectionStatusChanged = callback;
  }

  /// Registers the callback that fires when the desktop rejects a pairing.
  void onPairingRequired(void Function(PairingChallenge challenge) callback) {
    _onPairingRequired = callback;
  }

  /// Registers the callback that fires when the desktop accepts a pairing and
  /// issues a session token.
  void onPairingAccepted(void Function(String sessionToken) callback) {
    _onPairingAccepted = callback;
  }

  /// Sets the session token issued after a successful pairing. Heartbeats carry
  /// it so the desktop can resolve the sending device.
  void setSessionToken(String? token) {
    _sessionToken = token;
  }

  /// Dials [target] and switches to `connected` once the socket is ready.
  /// When no session token is known, the desktop cannot authorize input yet,
  /// so the status becomes `pairingRequired` and a PIN challenge fires instead
  /// of going straight to `connected`.
  Future<void> connect(ConnectionTarget target) async {
    _lastTarget = target;
    _reconnectTimer?.cancel();
    _reconnectTimer = null;
    _setStatus(ConnectionStatus.connecting);

    final uri = target.resolve(defaultPort: defaultPort);
    final channel = await _connect(uri);

    _channel = channel;
    _subscription = channel.stream.listen(
      _onMessage,
      onError: (_) => _handleConnectionClosed(),
      onDone: _handleConnectionClosed,
    );

    _startHeartbeat();
    _sendHeartbeat();
    if (_sessionToken == null) {
      _setStatus(ConnectionStatus.pairingRequired);
      _onPairingRequired?.call(
        const PairingChallenge(method: PairingMethod.pin),
      );
    } else {
      _setStatus(ConnectionStatus.connected);
    }
  }

  /// Closes the socket and returns to `disconnected`.
  Future<void> disconnect() async {
    _reconnectTimer?.cancel();
    _reconnectTimer = null;
    _heartbeatTimer?.cancel();
    _heartbeatTimer = null;
    _lastTarget = null;
    await _channel?.sink.close();
    _subscription = null;
    _channel = null;
    _setStatus(ConnectionStatus.disconnected);
  }

  /// Sends a `state` frame with an incrementing sequence number.
  void sendState({
    required double steering,
    required double accelerator,
    required double brake,
    required double clutch,
  }) {
    _send({
      'type': 'state',
      'seq': ++_seq,
      'steering': steering,
      'accelerator': accelerator,
      'brake': brake,
      'clutch': clutch,
    });
  }

  /// Sends a `button` frame using the wire values shared with the desktop.
  void sendButtonEvent(ControlId control, ActionType action) {
    _send({
      'type': 'button',
      'control': control.wireValue,
      'action': action.wireValue,
    });
  }

  /// Sends a `pair_request` for the current device.
  void submitPairingCode(String code) {
    _send({
      'type': 'pair_request',
      'device_id': deviceId,
      'code': code,
    });
  }

  /// Sends the desired dashboard input mapping (`keyboard` or `gamepad`).
  void sendMappingMode(InputMapping mapping) {
    _send({
      'type': 'mapping',
      'mode': mapping.wireValue,
    });
  }

  void _sendHeartbeat() {
    _send({
      'type': 'heartbeat',
      'session_token': ?_sessionToken,
    });
  }

  void _onMessage(dynamic data) {
    if (data is! String) {
      return;
    }

    final Object? decoded;
    try {
      decoded = jsonDecode(data);
    } on FormatException {
      return;
    }

    if (decoded is! Map<String, dynamic>) {
      return;
    }

    switch (decoded['type']) {
      case 'pair_response':
        final accepted = decoded['accepted'] == true;
        if (accepted) {
          final sessionToken = decoded['session_token'];
          if (sessionToken is String) {
            _sessionToken = sessionToken;
            _onPairingAccepted?.call(sessionToken);
          }

          _setStatus(ConnectionStatus.connected);
        } else {
          _setStatus(ConnectionStatus.pairingRequired);
          _onPairingRequired?.call(
            const PairingChallenge(method: PairingMethod.pin),
          );
        }
    }
  }

  void _handleConnectionClosed() {
    _heartbeatTimer?.cancel();
    _heartbeatTimer = null;
    _subscription?.cancel();
    _subscription = null;
    _channel = null;

    final target = _lastTarget;
    if (target == null) {
      _setStatus(ConnectionStatus.disconnected);
      return;
    }

    _setStatus(ConnectionStatus.reconnecting);
    _reconnectTimer?.cancel();
    _reconnectTimer = Timer(reconnectInterval, () {
      connect(target);
    });
  }

  void _startHeartbeat() {
    _heartbeatTimer?.cancel();
    _heartbeatTimer = Timer.periodic(heartbeatInterval, (_) {
      _sendHeartbeat();
    });
  }

  void _send(Map<String, dynamic> message) {
    final channel = _channel;
    if (channel == null) {
      return;
    }

    channel.sink.add(jsonEncode(message));
  }

  void _setStatus(ConnectionStatus next) {
    if (_status == next) {
      return;
    }

    _status = next;
    _onConnectionStatusChanged?.call(next);
  }

  static Future<StreamChannel<dynamic>> _defaultConnect(Uri uri) async {
    final channel = WebSocketChannel.connect(uri);
    await channel.ready;
    return channel;
  }
}

File: android/app/src/main/kotlin/dev/fazrigading/wheeldeck/data/services/WheelDeckClient.kt
package dev.fazrigading.wheeldeck.data.services

import android.util.Log
import dev.fazrigading.wheeldeck.domain.models.Button
import dev.fazrigading.wheeldeck.domain.models.ConnectionStatus
import dev.fazrigading.wheeldeck.domain.models.ConnectionTarget
import dev.fazrigading.wheeldeck.domain.models.DefaultWheelDeckPort
import dev.fazrigading.wheeldeck.domain.models.Heartbeat
import dev.fazrigading.wheeldeck.domain.models.Mapping
import dev.fazrigading.wheeldeck.domain.models.PairRequest
import dev.fazrigading.wheeldeck.domain.models.PairResponse
import dev.fazrigading.wheeldeck.domain.models.State
import dev.fazrigading.wheeldeck.domain.models.Unpair
import dev.fazrigading.wheeldeck.domain.models.WireJson
import dev.fazrigading.wheeldeck.domain.models.WireMessage
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString
import java.util.concurrent.atomic.AtomicLong

class WheelDeckClient(
    private val deviceId: String,
    private val heartbeatIntervalMs: Long = 2000L,
    private val reconnectIntervalMs: Long = 3000L,
    private val scope: CoroutineScope = CoroutineScope(Dispatchers.Default),
) : WebSocketListener() {
    private val client = OkHttpClient()

    private val _status = MutableStateFlow(ConnectionStatus.Disconnected)
    val status: StateFlow<ConnectionStatus> = _status.asStateFlow()

    private val seq = AtomicLong(0)
    private var webSocket: WebSocket? = null
    private var sessionToken: String? = null

    /// True between [disconnect] and the next [connect]; suppresses the error
    /// log for the socket teardown the user asked for.
    private var userRequestedClose = false

    /// Last connection target, used to reconnect after a lifecycle pause.
    var lastTarget: ConnectionTarget? = null
        private set
    private var heartbeatJob: Job? = null
    private var reconnectJob: Job? = null

    private var onConnectionStatusChanged: ((ConnectionStatus) -> Unit)? = null
    private var onPairingRequired: (() -> Unit)? = null
    private var onPairingAccepted: ((String) -> Unit)? = null
    private var onUnpair: (() -> Unit)? = null

    companion object {
        private const val TAG = "WheelDeckClient"
    }

    fun onConnectionStatusChanged(callback: (ConnectionStatus) -> Unit) {
        onConnectionStatusChanged = callback
    }

    fun onPairingRequired(callback: () -> Unit) {
        onPairingRequired = callback
    }

    fun onPairingAccepted(callback: (String) -> Unit) {
        onPairingAccepted = callback
    }

    fun onUnpair(callback: () -> Unit) {
        onUnpair = callback
    }

    fun setSessionToken(token: String?) {
        sessionToken = token
    }

    fun connect(target: ConnectionTarget) {
        lastTarget = target
        userRequestedClose = false
        cancelReconnectTimer()
        setStatus(ConnectionStatus.Connecting)

        val uri = target.resolve(defaultPort = DefaultWheelDeckPort)
        val request = Request.Builder()
            .url(uri)
            .build()

        webSocket = client.newWebSocket(request, this)
    }

    fun disconnect() {
        lastTarget = null
        userRequestedClose = true
        cancelReconnectTimer()
        cancelHeartbeatTimer()

        // Graceful close would wait on the close handshake and hang
        // MockWebServer-based tests; cancel() is proven and the unpair frame
        // is flushed with a short delay by the caller that needs it.
        webSocket?.cancel()
        webSocket = null
        setStatus(ConnectionStatus.Disconnected)
    }

    fun sendState(
        steering: Double,
        accelerator: Double,
        brake: Double,
        clutch: Double,
        cameraX: Double = 0.0,
        cameraY: Double = 0.0,
    ) {
        val message = State(
            seq = seq.incrementAndGet(),
            steering = steering,
            accelerator = accelerator,
            brake = brake,
            clutch = clutch,
            cameraX = cameraX,
            cameraY = cameraY,
        )
        sendMessage(message)
    }

    fun sendButtonEvent(control: String, action: String) {
        val message = Button(
            control = control,
            action = action,
        )
        sendMessage(message)
    }

    fun submitPairingCode(code: String) {
        val message = PairRequest(
            deviceId = deviceId,
            code = code,
        )
        sendMessage(message)
    }

    fun sendMappingMode(mode: String) {
        val message = Mapping(
            mode = mode,
        )
        sendMessage(message)
    }

    /// Tells the desktop to revoke this device's pairing. Send while the socket
    /// is still open, before disconnecting.
    fun sendUnpair() {
        sendMessage(Unpair)
    }

    override fun onOpen(webSocket: WebSocket, response: Response) {
        Log.d(TAG, "WebSocket opened")

        if (sessionToken == null) {
            setStatus(ConnectionStatus.PairingRequired)
            onPairingRequired?.invoke()
        } else {
            setStatus(ConnectionStatus.Connected)
        }

        scheduleHeartbeat()
        sendHeartbeat()
    }

    override fun onMessage(webSocket: WebSocket, text: String) {
        Log.d(TAG, "Received message: $text")
        handleTextMessage(text)
    }

    override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
        Log.d(TAG, "Received binary message: ${bytes.hex()}")
    }

    override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
        Log.d(TAG, "WebSocket closed: $code $reason")
        webSocketCleanup()
        handleConnectionClosed()
    }

    override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
        Log.d(TAG, "WebSocket closing: $code $reason")
        webSocket.close(code, reason)
    }

    override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
        when {
            userRequestedClose -> Log.d(TAG, "WebSocket closed by user: ${t.message}")
            // Already in the reconnect loop: a failed retry (e.g. ENETUNREACH
            // while Wi-Fi is still down) is expected, not an error.
            _status.value == ConnectionStatus.Reconnecting ->
                Log.d(TAG, "Reconnect attempt failed: ${t.message}")
            else -> Log.e(TAG, "WebSocket failure", t)
        }
        webSocketCleanup()

        // Connection errors like connection refused are reconnectable
        val isReconnectable = response == null
        if (isReconnectable) {
            handleConnectionClosed()
        } else {
            setStatus(ConnectionStatus.Disconnected)
        }
    }

    private fun handleTextMessage(text: String) {
        try {
            val message = WireJson.decodeFromString<WireMessage>(text)

            when (message) {
                is PairResponse -> {
                    if (message.accepted) {
                        message.sessionToken?.let { token ->
                            sessionToken = token
                            onPairingAccepted?.invoke(token)
                        }
                        setStatus(ConnectionStatus.Connected)
                    } else {
                        setStatus(ConnectionStatus.PairingRequired)
                        onPairingRequired?.invoke()
                    }
                }
                is Unpair -> {
                    Log.d(TAG, "Desktop revoked this pairing")
                    onUnpair?.invoke()
                }
                else -> {
                    // Other message types are handled by the UI layer
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse message: $text", e)
        }
    }

    private fun sendMessage(message: WireMessage) {
        val json = WireJson.encodeToString(message)
        webSocket?.send(json)
    }

    private fun scheduleHeartbeat() {
        heartbeatJob?.cancel()
        heartbeatJob = scope.launch {
            while (isActive) {
                delay(heartbeatIntervalMs)
                sendHeartbeat()
            }
        }
    }

    private fun cancelHeartbeatTimer() {
        heartbeatJob?.cancel()
        heartbeatJob = null
    }

    private fun sendHeartbeat() {
        if (webSocket == null) return

        val message = Heartbeat(
            sessionToken = sessionToken,
        )
        sendMessage(message)
    }

    private fun webSocketCleanup() {
        webSocket?.let {
            try {
                it.cancel()
            } catch (_: CancellationException) {
            } catch (_: Exception) {
            }
        }
        webSocket = null
    }

    private fun handleConnectionClosed() {
        cancelHeartbeatTimer()

        val target = lastTarget
        if (target == null) {
            setStatus(ConnectionStatus.Disconnected)
            return
        }

        setStatus(ConnectionStatus.Reconnecting)
        scheduleReconnect(target)
    }

    private fun scheduleReconnect(target: ConnectionTarget) {
        reconnectJob?.cancel()
        reconnectJob = scope.launch {
            delay(reconnectIntervalMs)
            connect(target)
        }
    }

    private fun cancelReconnectTimer() {
        reconnectJob?.cancel()
        reconnectJob = null
    }

    private fun setStatus(next: ConnectionStatus) {
        if (_status.value == next) return
        _status.value = next
        onConnectionStatusChanged?.invoke(next)
    }
}
